File: src/main/java/rjojjr/com/github/taskslock/TasksLockService.java
package rjojjr.com.github.taskslock;

import jakarta.annotation.PreDestroy;
import rjojjr.com.github.taskslock.models.TaskLock;

import java.util.List;

public interface TasksLockService {

    /**
     * Acquire lock for task
     * @param taskName unique task identifier
     * @param contextId a tracing identifier provided by the consumer
     * @param waitForLock block until lock is acquired
     * @return TaskLock object
     */
    TaskLock acquireLock(String taskName, String contextId, boolean waitForLock);

    /**
     * Release lock for task
     * @param taskName unique task identifier
     * @return contextId a tracing identifier provided by the consumer
     */
    String releaseLock(String taskName);

    /**
     * Release locks for all tasks
     * @return contextId a tracing identifier provided by the consumer
     */
    String releaseLocks();

    /**
     * Acquire lock with embedded impl.
     * @param taskName unique task identifier
     * @param hostName hostname of application/container acquiring lock
     * @param contextId a tracing identifier provided by the consumer
     * @param waitForLock block until lock is acquired
     * @return TaskLock object
     */
    TaskLock acquireLock(String taskName, String hostName, String contextId, boolean waitForLock);

    /**
     * List locks for all tasks
     * @param contextId a tracing identifier provided by the consumer
     * @return List<TaskLock> locks
     */
    List<TaskLock> getLocks(String contextId);

    /**
     * Release all locks & cleanup on shutdown
     */
    @PreDestroy
    void onDestroy();

}

File: client-modules/python/src/taskslock/client/tasks_lock_service.py
from utils import host_utils
import requests
import json


class TasksLockService:
    # TODO - 6IA - release all locks on shutdown
    def __init__(self,
                 protocol: str | None = None,
                 host: str | None = None,
                 port: int | None = None):
        self._protocol = protocol if not protocol is None else host_utils.get_host()
        self._host = host if not host is None else host_utils.get_host()
        self._port = port if not port is None else host_utils.get_port()
        self._url = f'{self._protocol}://{self._host}:{str(self._port)}'

    def acquire_lock(self, task_name: str, context_id: str, wait_for_lock: bool):
        response = requests.get(f'{self._url}/tasks-lock/api/v1/acquire?taskName={task_name}&contextId={context_id}&waitForLock={"true" if wait_for_lock else "false"}')
        if response.status_code < 300:
            body = json.loads(response.content)
            return TaskLock(body, lambda: self.release_lock(task_name))

    def release_lock(self, task_name: str) -> bool:
        response = requests.get(f'{self._url}/tasks-lock/api/v1/release?taskName={task_name}')
        if response.status_code < 300:
            return True
        return False
